### PR TITLE
Add support for creating AMP pages

### LINKED ISSUE #6568: AMP - Add support for creating AMP pages
AMP pages should be opt-in following convention.

The current proposed behavior is to create a file in the `pages/` directory with the `.amp.js` extension.

If a page exists as both `.js` and `.amp.js`, it should be able to be rendered with or without AMP-mode (toggled via `?amp=1`).

### PR DESCRIPTION
**Pull request by ijjk**
Adds support for opting into AMP pages using `.amp.js`. When both `dog.js` and `dog.amp.js` exist the user has to specify `?amp=1` to render `dog.amp.js`. If only `dog.amp.js` then `/dog` renders to that. If `/dog?amp=1` is called and there is no `dog.amp.js` it results in a 404

Closes: #6568

### CODE REVIEW

File: packages/next/server/on-demand-entry-handler.js [hunk 1]
```diff
@@ -232,6 +232,10 @@ export default function onDemandEntryHandler (devMiddleware, multiCompiler, {
       const name = join('static', buildId, 'pages', bundleFile)
       const absolutePagePath = pagePath.startsWith('next/dist/pages') ? require.resolve(pagePath) : join(pagesDir, pagePath)
 
+      if (pagePath.indexOf('index.amp') > -1) {
+        page = pagePath.replace('.js', '')
```
Comment (Timer) [inline packages/next/server/on-demand-entry-handler.js]: This should probably be consolidated, considering we already replace via ``new RegExp(`\\.+(?:${pageExtensions.join('|')})$`)`` a few lines above.

File: examples/experimental-amp/pages/cat/index.amp.js [hunk 2]
```diff
@@ -0,0 +1 @@
+export { default } from './index'
```
Comment (Timer) [inline examples/experimental-amp/pages/cat/index.amp.js]: Let's adjust these examples to show just `.js`, just `.amp.js` and then a combination of both for a single page?

**Review by Timer — CHANGES_REQUESTED** [review]

File: examples/experimental-amp/pages/cat/index.amp.js (hunk 2 above)
Comment (ijjk) [reply examples/experimental-amp/pages/cat/index.amp.js]: Updated

File: packages/next/server/on-demand-entry-handler.js (hunk 1 above)
Comment (ijjk) [reply packages/next/server/on-demand-entry-handler.js]: Yeah, it should be more consolidated now

File: packages/next/server/on-demand-entry-handler.js [hunk 3]
```diff
@@ -232,6 +233,9 @@ export default function onDemandEntryHandler (devMiddleware, multiCompiler, {
       const name = join('static', buildId, 'pages', bundleFile)
       const absolutePagePath = pagePath.startsWith('next/dist/pages') ? require.resolve(pagePath) : join(pagesDir, pagePath)
 
+      page = pageUrl.replace('//', '/')
```
Comment (Timer) [inline packages/next/server/on-demand-entry-handler.js]: Let's use the normalize helper or `path.posix.normalize()`.

File: packages/next/server/on-demand-entry-handler.js [hunk 4]
```diff
@@ -215,7 +215,8 @@ export default function onDemandEntryHandler (devMiddleware, multiCompiler, {
         throw pageNotFoundError(normalizedPagePath)
       }
 
-      let pagePath = await findPageFile(pagesDir, normalizedPagePath, pageExtensions)
+      let pagePath = await findPageFile(pagesDir, normalizedPagePath, pageExtensions, amp)
+      const isAmp = pagePath && pagePath.endsWith('amp.js')
```
Comment (Timer) [inline packages/next/server/on-demand-entry-handler.js]: This needs to work with all `pageExtensions`, not just `.js`.

File: packages/next/server/lib/find-page-file.ts [hunk 5]
```diff
@@ -1,8 +1,12 @@
 import { join } from 'path'
 import {isWriteable} from '../../build/is-writeable'
 
-export async function findPageFile(rootDir: string, normalizedPagePath: string, pageExtensions: string[]): Promise<string|null> {
-  for (const extension of pageExtensions) {
+export async function findPageFile(rootDir: string, normalizedPagePath: string, pageExtensions: string[], amp: boolean): Promise<string|null> {
+  // Add falling back to .amp.js extension
+  if (!amp) pageExtensions = pageExtensions.concat('amp.js')
```
Comment (Timer) [inline packages/next/server/lib/find-page-file.ts]: This needs to work with all `pageExtensions`, not just `.js`.

**Review by Timer — CHANGES_REQUESTED** [review]

File: packages/next/server/lib/find-page-file.ts (hunk 5 above)
Comment (ijjk) [reply packages/next/server/lib/find-page-file.ts]: Good catch

**Review by Timer — APPROVED** [review]

**Comment by Timer** [comment]
This is beautiful work, @ijjk.

**Pull request merged**